drivers/mbox/mbox_nrf_bellboard_local.c

```diff
@@ -0,0 +1,169 @@
+
+BUILD_ASSERT((DT_INST_PROP_LEN(0, nordic_interrupt_mapping) % 2) == 0,
+	     "# interrupt mappings not specified in pairs");
+
+/* BELLBOARD event mappings */
+#define EVT_MAPPING_ITEM(idx) DT_INST_PROP_BY_IDX(0, nordic_interrupt_mapping, idx)
+#define BELLBOARD_GET_EVT_MAPPING(idx, _)                                                          \
+	COND_CODE_1(                                                                               \
+		DT_INST_PROP_HAS_IDX(0, nordic_interrupt_mapping, UTIL_INC(UTIL_X2(idx))),         \
+		([EVT_MAPPING_ITEM(UTIL_INC(UTIL_X2(idx)))] = EVT_MAPPING_ITEM(UTIL_X2(idx)),),    \
+		())
+
+static const uint32_t evt_mappings[BELLBOARD_NUM_IRQS] = {
+	LISTIFY(DT_NUM_IRQS(DT_DRV_INST(0)), BELLBOARD_GET_EVT_MAPPING, ())};
+
+/* BELLBOARD instance */
+static NRF_BELLBOARD_Type *bellboard = (NRF_BELLBOARD_Type *)DT_INST_REG_ADDR(0);
+
+/* BELLBOARD runtime resources */
+static mbox_callback_t cbs[NRF_BELLBOARD_EVENTS_TRIGGERED_COUNT];
+static void *cbs_ctx[NRF_BELLBOARD_EVENTS_TRIGGERED_COUNT];
+static uint32_t evt_enabled_masks[BELLBOARD_NUM_IRQS];
+
+static void bellboard_local_isr(const void *parameter)
+{
+	uint8_t irq_idx = (uint8_t)(uintptr_t)parameter;
+	uint32_t int_pend;
+
+	int_pend = nrf_bellboard_int_pending_get(bellboard, irq_idx);
+
+	for (uint8_t i = 0U; i < NRF_BELLBOARD_EVENTS_TRIGGERED_COUNT; i++) {
+		nrf_bellboard_event_t event = nrf_bellboard_triggered_event_get(i);
+
+		if (nrf_bellboard_event_check(bellboard, event)) {
+			nrf_bellboard_event_clear(bellboard, event);
+		}
+
+		if ((int_pend & BIT(i)) != 0U) {
+			if (cbs[i] != NULL) {
+				cbs[i](DEVICE_DT_INST_GET(0), i, cbs_ctx[i], NULL);
+			}
+		}
+	}
+}
+
+static uint32_t bellboard_local_max_channels_get(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return NRF_BELLBOARD_EVENTS_TRIGGERED_COUNT;
+}
+
+static int bellboard_local_register_callback(const struct device *dev, uint32_t id,
+					     mbox_callback_t cb, void *user_data)
+{
+	ARG_UNUSED(dev);
+
+	if (id >= NRF_BELLBOARD_EVENTS_TRIGGERED_COUNT) {
+		return -EINVAL;
+	}
+
+	cbs[id] = cb;
+	cbs_ctx[id] = user_data;
+
+	return 0;
+}
+
+static int bellboard_local_set_enabled(const struct device *dev, uint32_t id, bool enable)
+{
+	bool valid_found = false;
+
+	ARG_UNUSED(dev);
+
+	if (id >= NRF_BELLBOARD_EVENTS_TRIGGERED_COUNT) {
+		return -EINVAL;
+	}
+
+	for (uint8_t i = 0U; i < BELLBOARD_NUM_IRQS; i++) {
+		uint32_t *evt_enabled_mask;
+
+		if ((evt_mappings[i] == 0U) || ((evt_mappings[i] & BIT(id)) == 0U)) {
+			continue;
+		}
+
+		valid_found = true;
+		evt_enabled_mask = &evt_enabled_masks[i];
+
+		if (enable) {
+			if ((*evt_enabled_mask & BIT(id)) != 0U) {
+				return -EALREADY;
+			}
+
+			*evt_enabled_mask |= BIT(id);
+			nrf_bellboard_int_enable(bellboard, i, BIT(id));
+		} else {
+			if ((*evt_enabled_mask & BIT(id)) == 0U) {
+				return -EALREADY;
+			}
+
+			*evt_enabled_mask &= ~BIT(id);
+			nrf_bellboard_int_disable(bellboard, i, BIT(id));
```

Comment:
Similarly to VEVIF, I don't think interrupts should get disabled unconditionally, because other still active channels may be assigned to the same interrupt.

Reply:
Note that here we’re never disabling IRQ, but changing the interrupt enabled masks.